### ISSUE TITLE
ES6 target / ESM distribution

### ISSUE THREAD

**Issue by alessiostalla** [vitaly-t/iter-ops#73]
Hi, would it be possible to release a version of the library compiled to ES6, so using and exporting native modules?
Recent versions of Angular complain about non-ESM dependencies (i.e. AMD or CommonJS).

**Comment by alessiostalla** [comment]
Yeah I know I can use the library in an Angular project and safely ignore that warning, but since our customer will also see the warning, they may, in turn, complain ;) 
What people usually do in such cases in the Java world is to generate multiple artifacts from the same project, such as, foo-with-bar-v1.0.0.jar, foo-no-bar-v1.0.0.jar. I don't know the convention in JS/TS and I have no idea how hard npmjs makes this (i.e. to publish multiple variants of the same package).

**Comment by vitaly-t** [comment]
I recall actually trying to add that ESM compliance, for another project, got stuck, even asked on StackOverflow about it (will link the question, if I find it), and there was something about Angular not a real ESM-compliance or something, I do not quite remember right now.

I will have another look at it later on. But if you want to try it yourself in the meantime, you're most welcome!

**Comment by vitaly-t** [comment]
@alessiostalla I wrote wrong before, the library's distribution comes with `ES2015`, which is equivalent for `ES6`.

![image](https://user-images.githubusercontent.com/5108906/146948123-3b834f27-e753-412d-8cdc-6898c0deda1c.png)

**Comment by alessiostalla** [comment]
Oh, nice. I'll do more tests on my part, maybe I've misreported this.

**Comment by FaberVitale** [comment]
> @alessiostalla I wrote wrong before, the library's distribution comes with `ES2015`, which is `ES6` equivalent.
> 

This project is really cool but the lack of esm module version makes it hard to consider as a valid alternative to rxjs in a client-side project because it is not tree-shakable.

Tree-shaking, aka the removal of unused code, is not possible for the following reasons:

1. This project does not expose an esm module version
2. This project that does not have `side-effects: false` in `package.json`

Please refer to the webpack docs for more details: https://webpack.js.org/guides/tree-shaking/#conclusion

I could submit a PR, If you want help with this issue.

**Comment by vitaly-t** [comment]
> I could submit a PR, If you want help with this issue.

Would love one, thank you!!!

**Comment by vitaly-t** [comment]
Thanks to @RebeccaStevens, this issue has been pretty much resolved, pending some extra documentation + tests on my end.

**Comment by RebeccaStevens** [comment]
The `sideEffects` field in package json hasn't been set to false yet.

**Comment by vitaly-t** [comment]
* `sideEffects` added
* circular dependencies resolved

Looks almost ready :)

**Comment by FaberVitale** [comment]
> * `sideEffects` added
> 
>     * circular dependencies resolved
> 
> 
> Looks almost ready :)


Adding the "exports" in package.json is a **breaking** change!
You should bump version to `2.x.x` before releasing a new version.

### Node std warning

```
Warning: Introducing the "exports" field prevents consumers of a package from using any entry points that are not defined, including the package.json (e.g. require('your-package/package.json'). This will likely be a breaking change.
```

source: https://nodejs.org/api/packages.html#package-entry-points


See also

- https://nodejs.org/api/packages.html#determining-module-system
- https://github.com/then/is-promise/issues/19

**Comment by vitaly-t** [comment]
@FaberVitale Fair enough, this is something to consider before going with the official release :)

In the meantime, I've just wrapped this all up and published a `1.5.0-beta.0`

```
npm i iter-ops@1.5.0-beta.0
```

or perhaps just:

```
npm i iter-ops@beta
```

Please give it a test, and let me know ;) My own tests with Angular were all pass ;)

**Comment by FaberVitale** [comment]
### bundlephobia

<img width="801" alt="Screenshot 2021-12-22 at 16 31 16" src="https://user-images.githubusercontent.com/16583125/147116440-12ada94c-2df6-41c4-8265-5314925fe127.png">

link: https://bundlephobia.com/package/iter-ops@1.5.0-beta.0

**Comment by vitaly-t** [comment]
Looks like we did do something awesome 😸

**Comment by vitaly-t** [comment]
And if we go for broke, and include absolutely everything, it is still a very manageable 4kb 😄 

![image](https://user-images.githubusercontent.com/5108906/147124357-7a99757e-d8ae-4ab5-8c71-e29cb64ab613.png)

**Comment by vitaly-t** [comment]
> Warning: Introducing the "exports" field prevents consumers of a package from using any entry points that are not defined, including the package.json (e.g. require('your-package/package.json'). This will likely be a breaking change.

I wouldn't be paying much attention to this. After all, this whole package is still very new anyhow. I think we will be fine going with 1.5.0 when it is finalized.

**Comment by vitaly-t** [comment]
Version 1.5.0 has been released. This issue is no more.

**Comment by vitaly-t** [comment]
Just so, 1.5.0 missed some declarations `.d.ts` that were patched in v1.5.1

@FaberVitale Were you ok using it?
